### ISSUE TITLE
Pipeline launched from main process waits for a background process

### ISSUE THREAD

**Issue by mil** [oilshell/oil#1002]
I'm attempting to use subshells in osh and it seems like after backgrounding a subshell, then all pipes (in the primary process) will deadlock.

```sh
#!/usr/bin/env osh
trap term TERM
trap term INT

term() { kill 0; exit 1; }
main() {
  {
    while true; do
      sleep 1
    done 
  } >/dev/null &

  echo 1
  echo foo | grep foo
  echo 2
  echo bar | grep bar
  echo 3
  echo baz | grep baz

  term
}

main
```

Produces the output (and then deadlocks):
```
1
foo
```

With `ash`/`bash` things work normally and I get:
```
1
foo
2
bar
3
baz
```

**Comment by mil** [comment]
Changing subshell body to terminate eventually confirms this is indeed a deadlock:

```sh
#!/usr/bin/env osh
trap term TERM
trap term INT

term() { kill 0; exit 1; }
main() {
  {
      sleep 10
  } >/dev/null &

  echo 1
  echo foo | grep foo
  echo 2
  echo bar | grep bar
  echo 3
  echo baz | grep baz

  term
}

main
```

For whatever reason the main process is waiting the subshell to terminate before allowing the pipe in the main process to go through.

**Comment by andychu** [comment]
Thanks for the report!    I'm still understanding this, but woah I get a segfault in both bash 4.4 and dash (on my Ubuntu system)?  

Are you seeing that?

zsh can run it, but hangs at the end.


```
$ bash demo/issue-1002.sh 
1
foo
2
bar
3
baz


Segmentation fault (core dumped)

```

**Comment by andychu** [comment]
Oh the crash goes away if I remove the `trap term TERM` line.  I think that is triggering an infinite loop or something like that!

OK that's a different issue.

Confirmed that bin/osh is behaving differently!

**Comment by andychu** [comment]
Ah OK it's the current process group, I guess both /bin/kill and builtin kill implement that

https://superuser.com/questions/563972/what-does-kill-0-do-actually

**Comment by andychu** [comment]
Hm somehow this works with option group `oil:basic` !  That's very surprising

```
~/git/oilshell/oil$ bin/osh demo/issue-1002.sh 
1
foo
...
...
...
...
...
...
^C
```

then

```

~/git/oilshell/oil$ bin/osh -O oil:basic demo/issue-1002.sh 
1
foo
...
2
bar
3
baz
Terminated

```

**Comment by andychu** [comment]
More info:

- Bug appears in version 0.6.0, from 7/2019, so it's quite old
- Oil 0.5.0 from 7/2018 runs it correctly! 

I could bisect it a bit more ...

**Comment by andychu** [comment]
OK I bisected it a bit more, and 0.6.pre4 from 9/2018 runs the script correctly, but 0.6.pre5 from 10/2018 doesn't!

http://www.oilshell.org/releases.html

There was a release announcement:

http://www.oilshell.org/blog/2018/10/11.html

Looking into it more ...

**Comment by andychu** [comment]
OK I found what's almost certainly the culprit!  

That release introduced `shopt -s lastpipe` semantics by default in OSH.

https://github.com/oilshell/oil/commit/5733f31bf4b878f5063b244498787f4820108342

Somehow that is buggy.  (`bash -O lastpipe` doesn't exhibit the bug, and zsh has `lastpipe` semantics by default and doesn't exhibit the bug)

**Comment by mil** [comment]
Nice to hear you found the root cause. Surprising it dates back to 2018, I'd consider this issue as a pretty common use case. Long running background threads are a pretty common pattern in my shell scripting.

**Comment by andychu** [comment]
OK this is fixed and will be out with the next release!  Great report.

I think this bug persisted because unless you have an infinite background process, the script will still "work", it will just be blocked unnecessarily.

I think it's not technically a deadlock because that is when two processes wait on each other?  I think the bug here is that one process is just waiting unnecessarily.

The core issue is that all the process done notifications are are asynchronous and "serialized" in the main process.  In the lastpipe case, we weren't updating the pipeline state, because the last one is a special case.  So the wait loop for a pipeline would basically do an extra turn and wait for a completely unrelated process!

----

(BTW I noticed some incomplete stuff in job control while doing this (#360).  And I also improved the `jobs` builtin listing, but it still is incomplete as well.  I'm not sure how vital this is; the basics work)
